File: Linked List, Hash/leetcode/21_merge_two_sorted_lists/21_merge_two_sorted_lists.cpp
#include <iostream>
#include <vector>

/* 
2개의 sorted linked list(non-decreasing order)가 존재할 때, 두 list를 merge하는 문제
linked list의 head를 return 해야 함

ex. 
list1 : [1, 2, 4]
list2 : [1, 3, 4]
output : [1, 1, 2, 3, 4, 4]
*/

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode sentinel;
        ListNode* curNode = &sentinel;

        while (list1 && list2) {
            if (list1->val <= list2->val) {
                curNode->next = list1;
                list1 = list1->next;
            } else {
                curNode->next = list2;
                list2 = list2->next;
            }
            curNode = curNode->next;
        }

        // Append the remaining nodes
        curNode->next = list1 ? list1 : list2;

        return sentinel.next;
    }
};

// Helper function to print the list
void printList(ListNode* node) {
    while (node != nullptr) {
        std::cout << node->val << " ";
        node = node->next;
    }
    std::cout << std::endl;
}

// Helper function to create a linked list from an array
ListNode* createList(const std::vector<int>& values) {
    ListNode* head = nullptr;
    ListNode* tail = nullptr;
    for (int value : values) {
        ListNode* newNode = new ListNode(value);
        if (!head) {
            head = newNode;
            tail = head;
        } else {
            tail->next = newNode;
            tail = tail->next;
        }
    }
    return head;
}

int main() {
    // Example lists to merge
    std::vector<int> list1_values = {1, 2, 4};
    std::vector<int> list2_values = {1, 3, 4};

    // Create the linked lists
    ListNode* list1 = createList(list1_values);
    ListNode* list2 = createList(list2_values);

    // Merge the lists
    Solution solution;
    ListNode* mergedList = solution.mergeTwoLists(list1, list2);

    // Print the merged list
    printList(mergedList);

    return 0;
}

File: Linked List, Hash/leetcode/21_merge_two_sorted_lists/21_merge_two_sorted_lists.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution1(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """

        sentinel = ListNode()
        curNode = sentinel

        while (list1 and list2): 
            if list1.val <= list2.val:
                curNode.next = list1
                list1 = list1.next
            else :
                curNode.next = list2
                list2 = list2.next
            curNode = curNode.next
        
        ## Append the remaining nodes
        if list1 :
            curNode.next = list1
        else :
            curNode.next = list2
        
        return sentinel.next

        
class Solution2(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        if list1 is None and list2 is not None:
            return list2
        elif list1 is not None and list2 is None:
            return list1
        head=None 
        prev=None
        while list1 is not None and list2 is not None:
            if list1.val<=list2.val:
                if head==None:
                    head=list1
                    prev=list1
                    list1=list1.next
                    continue
                prev.next=list1
                prev=list1
                list1=list1.next
                
            elif list1.val>list2.val:
                if head==None:
                    head=list2
                    prev=list2
                    list2=list2.next
                    continue
                prev.next=list2
                prev=list2
                list2=list2.next
        if prev==None:
            return head
        if list1 is not None:
            prev.next=list1
        else:
            prev.next=list2
        return head
        
class Solution3(object): ## 재귀적으로 푸는 방법 추가
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        # 기본 케이스: 하나의 리스트가 비어 있으면 다른 리스트 반환
        if not list1:
            return list2
        if not list2:
            return list1
        
        # 재귀적으로 두 리스트를 합병
        if list1.val < list2.val:
            list1.next = self.mergeTwoLists(list1.next, list2)
            return list1
        else:
            list2.next = self.mergeTwoLists(list1, list2.next)
            return list2

